### PR TITLE
Add support of Grain input pipeline for DPO.

### PR DESCRIPTION
**Pull request by igorts-git**
# Description

Add support of Grain input reading when using the new Tunix-based DPO/ORPO.
Modify `dpo.yml` to have some reasonable defaults that allow invoking DPO without extra config parameters.

Correct the case where `config.dpo.max_prompt_length` is not set. The default value is `max_target_length // 2`. It should be the same value that is passed to both the input pipeline and to the Tunix DPOTrainer class. Thus, I moved its computation to `types.py`.

In a follow up PR I will add a detailed logits comparison test for DPO.

BUGS: b/485626968

# Tests

CI tests.
Ran DPO/ORPO while reading using Grain from parquet files.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/input_pipeline/grain_data_processing.py [hunk 1]
```diff
@@ -263,6 +264,46 @@ def pretrain_preprocessing_pipeline(
...
+          pad_id=pad_id,
+          max_target_length=config.max_target_length,
+          data_column_names=data_columns,
+          max_prompt_length=max_prompt_length,
+      )
+  )
+
+  batch_size = config.global_batch_size_to_load // jax.process_count()
+  if config.grain_use_elastic_iterator:
+    # ElasticIterator batches internally, so return the pre-batch dataset.
```
[the remark this answers is not in the document]
Comment (igorts-git) [reply src/maxtext/input_pipeline/grain_data_processing.py:294]: done

File: src/maxtext/input_pipeline/dpo_utils.py [hunk 2]
```diff
@@ -31,6 +31,10 @@ class DPODataFormatting(grain.MapTransform):
   data_column_names: tuple[str, ...]
   max_prompt_length: int | None = None
 
+  def __post_init__(self):
+    if self.max_prompt_length is None:
```
Comment (igorts-git) [inline src/maxtext/input_pipeline/dpo_utils.py]: The new logic that I added in `types.py` guarantees that this value is not None. However, for cases like unit tests it is still useful to have this default computed. Let' me know if you want it removed.

**Review by aireenmei — APPROVED** [review]
LGTM. Thanks!

**Review by abhinavclemson — APPROVED** [review]

File: src/maxtext/input_pipeline/dpo_utils.py (hunk 2 above)
Comment (igorts-git) [reply src/maxtext/input_pipeline/dpo_utils.py]: Per private chat with Abhinav, I removed this part and updated the test to explicitly set `max_prompt_length`
